Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp

```diff
@@ -0,0 +1,700 @@
+        std::memmove(&tcp_buf.data[0x36], &tcp_buf.data[0x36 + ack_size], new_data_size);
+        tcp_buf.data_size -= ack_size;
+        tcp_buf.seq_id += ack_size;
+        tcppart->sequence_number = htonl(tcp_buf.seq_id);
+        Common::IPv4Header* ippart = (Common::IPv4Header*)&tcp_buf.data[14];
+        ippart->total_len = htons(tcp_buf.data_size - 14);
+        tcppart->checksum = 0;
+        tcppart->checksum = Common::ComputeTCPNetworkChecksum(ref->from, ref->to, tcppart,
+                                                              new_data_size + 20, IPPROTO_TCP);
+      }
+    }
+  }
+}
+
+/// <summary>
+/// This is a litle hack, Mario Kart open some UDP port
+/// and listen to it. We open it on our side manualy.
+/// </summary>
+void CEXIETHERNET::BuiltInBBAInterface::InitUDPPort(u16 port)
+{
+  StackRef* ref = GetAvaibleSlot(htons(port));
+  if (ref == nullptr || ref->ip != 0)
+    return;
+  ref->ip = 0x08080808;  // change for ip
+  ref->local = htons(port);
+  ref->remote = htons(port);
+  ref->type = 17;
+  ref->bba_mac = *(Common::MACAddress*)&m_eth_ref->mBbaMem[BBA_NAFR_PAR0];
+  ref->my_mac = fake_mac;
+  ref->from.sin_addr.s_addr = 0;
+  ref->from.sin_port = htons(port);
+  ref->to.sin_addr.s_addr = m_current_ip;
+  ref->to.sin_port = htons(port);
+  ref->udp_socket.setBlocking(false);
+  if (ref->udp_socket.bind(port) != sf::Socket::Done)
+  {
+    ERROR_LOG_FMT(SP1, "Couldn't open UDP socket");
+    return;
+  }
+}
+
+void CEXIETHERNET::BuiltInBBAInterface::HandleUDPFrame(Common::EthernetHeader* hwdata,
+                                                       Common::IPv4Header* ipdata,
+                                                       Common::UDPHeader* udpdata, u8* data)
+{
+  sf::IpAddress target;
+
+  if (*(u32*)ipdata->destination_addr == 0)
+    *(u32*)ipdata->destination_addr = m_router_ip;
+  // dns request
+  StackRef* ref = GetAvaibleSlot(udpdata->source_port);
+  if (ref->ip == 0)
+  {
+    ref->ip = *(u32*)ipdata->destination_addr;  // change for ip
+    ref->local = udpdata->source_port;
+    ref->remote = udpdata->destination_port;
+    ref->type = 17;
+    ref->bba_mac = *(Common::MACAddress*)&m_eth_ref->mBbaMem[BBA_NAFR_PAR0];
+    ref->my_mac = fake_mac;
+    ref->from.sin_addr.s_addr = *(u32*)&ipdata->destination_addr;
+    ref->from.sin_port = udpdata->destination_port;
+    ref->to.sin_addr.s_addr = *(u32*)&ipdata->source_addr;
+    ref->to.sin_port = udpdata->source_port;
+    ref->udp_socket.setBlocking(false);
+    if (ref->udp_socket.bind(htons(udpdata->source_port)) != sf::Socket::Done &&
+        ref->udp_socket.bind(sf::Socket::AnyPort) != sf::Socket::Done)
+    {
+      ERROR_LOG_FMT(SP1, "Couldn't open UDP socket");
+      return;
+    }
+  }
+  if (ntohs(udpdata->destination_port) == 1900)
+  {
+    InitUDPPort(26512);                                 // MK DD and 1080
+    InitUDPPort(26502);                                 // Air Ride
```

Comment:
Shouldn't you check/handle if these succeeded? Also, does it make sense to always open both of these?

Reply:
they are opened only if port 1900 is used, there is no way to tell the game playing from the BBA point of view. This is also a hack to facilitate the communication, it's not 100% required to work. if you have a way to popup a warning without killing dolphin i would add a check. for now only an error log is displayed

Reply:
I think you want `PanicAlertFmtT()`.